### DISCUSSION TITLE
Extensibility of xunit core framework components

### DISCUSSION THREAD

**Discussion by maurei** [xunit/xunit#2273] [Question] [answered]
Hi folks,

I'm wondering if it is possible to run my tests with a customised `XunitTestClassRunner` without doing a lot of heavy surgery. I would like to play around with overriding [`CreateClassFixture`](https://github.com/xunit/xunit/blob/fe6472890284faada2c346aa2b984429a3babf49/src/xunit.v3.core/Sdk/v3/Runners/XunitTestClassRunner.cs#L74).

I noticed that components like these are [instantiated manually](https://github.com/xunit/xunit/blob/c54cc52ffb275c81afed022521870193bbca6c39/src/xunit.execution/Sdk/Frameworks/Runners/XunitTestAssemblyRunner.cs#L235), so I guess extensibility using DI or so is off the table.

Still, the places where that happens are all `virtual`, so perhaps it could be done. I just haven't figured out how to hook into it.

Thanks

**Comment by bradwilson** [comment, marked as answer]
Heavy lifting is all we can offer today, I'm afraid. You are correct that there is no DI anywhere in xUnit.net v2.

We have an [Observation framework sample](https://github.com/xunit/samples.xunit/tree/main/ObservationExample) which shows how to do the heavy lifting. It's not easy to get in the middle, but you can start to opt out once you get to the place you want to get. So for you that means you'll need an assembly runner, test collection runner, etc., until you get to where you want to go. The virtuals at least make this fairly straight forward, even if you do end up having to write a dozen classes. 🙁
